Enabling DNS retryOnTimeout with TCP

Netty sometimes get a 5000ms timeout error when querying DNS. e.g.
`Caused by: io.netty.resolver.dns.DnsNameResolverTimeoutException: [19509: /172.33.0.2:53] DefaultDnsQuestion(<hidden> IN A) query '19509' via UDP timed out after 5000 milliseconds (no stack trace available)`

The team has now merged this branch that allows for a retry over TCP instead of UDP, in case of timeout.
I do not think it is possible to set this boolean from reactor-netty?
I would expect it to be in the `NameResolverProvider.NameResolverSpec` 

The merged branch:
https://github.com/netty/netty/pull/13757

--- a/docs/asciidoc/http-client.adoc
+++ b/docs/asciidoc/http-client.adoc
@@ -564,6 +564,8 @@ Default: -1 (to determine the value from the OS on Unix or use a value of 1 othe
 | `queryTimeout` | Sets the timeout of each DNS query performed by this resolver (resolution: milliseconds). Default: 5000.
 | `resolveCache` | The cache to use to store resolved DNS entries.
 | `resolvedAddressTypes` | The list of the protocol families of the resolved address.
+| `retryTcpOnTimeout` | Specifies whether this resolver will also fallback to TCP if a timeout is detected.
+By default, the resolver will only try to use TCP if the response is marked as truncated.
 | `roundRobinSelection` | Enables an
 {nettyjavadoc}/io/netty/resolver/AddressResolverGroup.html[`AddressResolverGroup`] of
 {nettyjavadoc}/io/netty/resolver/dns/DnsNameResolver.html[`DnsNameResolver`] that supports random selection

--- a/docs/asciidoc/tcp-client.adoc
+++ b/docs/asciidoc/tcp-client.adoc
@@ -418,6 +418,8 @@ Default: {nettyjavadoc}/io/netty/resolver/DefaultHostsFileEntriesResolver.html[`
 | `queryTimeout` | Sets the timeout of each DNS query performed by this resolver (resolution: milliseconds). Default: 5000.
 | `resolveCache` | The cache to use to store resolved DNS entries.
 | `resolvedAddressTypes` | The list of the protocol families of the resolved address.
+| `retryTcpOnTimeout` | Specifies whether this resolver will also fallback to TCP if a timeout is detected.
+By default, the resolver will only try to use TCP if the response is marked as truncated.
 | `roundRobinSelection` | Enables an
  {nettyjavadoc}/io/netty/resolver/AddressResolverGroup.html[`AddressResolverGroup`] of
  {nettyjavadoc}/io/netty/resolver/dns/DnsNameResolver.html[`DnsNameResolver`] that supports random selection

--- a/reactor-netty-core/build.gradle
+++ b/reactor-netty-core/build.gradle
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2023 VMware, Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2020-2024 VMware, Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -240,6 +240,8 @@ task japicmp(type: JapicmpTask) {
 	compatibilityChangeExcludes = [ "METHOD_NEW_DEFAULT" ]
 
 	methodExcludes = [
+			'reactor.netty.transport.NameResolverProvider#retryTcpOnTimeout()',
+			'reactor.netty.transport.NameResolverProvider$NameResolverSpec#retryTcpOnTimeout(boolean)'
 	]
 }
 

--- a/reactor-netty-core/src/main/java/reactor/netty/transport/NameResolverProvider.java
+++ b/reactor-netty-core/src/main/java/reactor/netty/transport/NameResolverProvider.java
@@ -211,6 +211,17 @@ public final class NameResolverProvider {
 		NameResolverSpec resolvedAddressTypes(ResolvedAddressTypes resolvedAddressTypes);
 
 		/**
+		 * Specifies whether this resolver will also fallback to TCP if a timeout is detected.
+		 * By default, the resolver will only try to use TCP if the response is marked as truncated.
+		 *
+		 * @param enable if true this resolver will also fallback to TCP if a timeout is detected,
+		 * if false the resolver will only try to use TCP if the response is marked as truncated.
+		 * @return {@code this}
+		 * @since 1.1.17
+		 */
+		NameResolverSpec retryTcpOnTimeout(boolean enable);
+
+		/**
 		 * Enables an {@link AddressResolverGroup} of {@link DnsNameResolver}s that supports random selection
 		 * of destination addresses if multiple are provided by the nameserver.
 		 * See {@link RoundRobinDnsAddressResolverGroup}.
@@ -379,6 +390,16 @@ public final class NameResolverProvider {
 	}
 
 	/**
+	 * Returns {@code true} if the resolver will also fallback to TCP if a timeout is detected.
+	 *
+	 * @return {@code true} if the resolver will also fallback to TCP if a timeout is detected
+	 * @since 1.1.17
+	 */
+	public boolean isRetryTcpOnTimeout() {
+		return retryTcpOnTimeout;
+	}
+
+	/**
 	 * Returns true if {@link RoundRobinDnsAddressResolverGroup} is in use.
 	 *
 	 * @return true if {@link RoundRobinDnsAddressResolverGroup} is in use
@@ -491,6 +512,7 @@ public final class NameResolverProvider {
 				queryTimeout.equals(that.queryTimeout) &&
 				Objects.equals(resolveCache, that.resolveCache) &&
 				resolvedAddressTypes == that.resolvedAddressTypes &&
+				retryTcpOnTimeout == that.retryTcpOnTimeout &&
 				roundRobinSelection == that.roundRobinSelection &&
 				// searchDomains is List so Objects.equals is OK
 				Objects.equals(searchDomains, that.searchDomains);
@@ -516,6 +538,7 @@ public final class NameResolverProvider {
 		result = 31 * result + Objects.hashCode(queryTimeout);
 		result = 31 * result + Objects.hashCode(resolveCache);
 		result = 31 * result + Objects.hashCode(resolvedAddressTypes);
+		result = 31 * result + Boolean.hashCode(retryTcpOnTimeout);
 		result = 31 * result + Boolean.hashCode(roundRobinSelection);
 		result = 31 * result + Objects.hashCode(searchDomains);
 		return result;
@@ -551,7 +574,7 @@ public final class NameResolverProvider {
 				.queryTimeoutMillis(queryTimeout.toMillis())
 				.eventLoop(group.next())
 				.channelFactory(() -> loop.onChannel(DatagramChannel.class, group))
-				.socketChannelFactory(() -> loop.onChannel(SocketChannel.class, group));
+				.socketChannelFactory(() -> loop.onChannel(SocketChannel.class, group), retryTcpOnTimeout);
 		if (bindAddressSupplier != null) {
 			// There is no check for bindAddressSupplier.get() == null
 			// This is deliberate, when null value is provided Netty will use the default behaviour
@@ -596,6 +619,7 @@ public final class NameResolverProvider {
 	final Duration queryTimeout;
 	final DnsCache resolveCache;
 	final ResolvedAddressTypes resolvedAddressTypes;
+	final boolean retryTcpOnTimeout;
 	final boolean roundRobinSelection;
 	final Iterable<String> searchDomains;
 
@@ -618,6 +642,7 @@ public final class NameResolverProvider {
 		this.queryTimeout = build.queryTimeout;
 		this.resolveCache = build.resolveCache;
 		this.resolvedAddressTypes = build.resolvedAddressTypes;
+		this.retryTcpOnTimeout = build.retryTcpOnTimeout;
 		this.roundRobinSelection = build.roundRobinSelection;
 		this.searchDomains = build.searchDomains;
 	}
@@ -650,6 +675,7 @@ public final class NameResolverProvider {
 		Duration queryTimeout = DEFAULT_QUERY_TIMEOUT;
 		DnsCache resolveCache;
 		ResolvedAddressTypes resolvedAddressTypes;
+		boolean retryTcpOnTimeout;
 		boolean roundRobinSelection;
 		Iterable<String> searchDomains;
 
@@ -752,6 +778,12 @@ public final class NameResolverProvider {
 		@Override
 		public NameResolverSpec resolvedAddressTypes(ResolvedAddressTypes resolvedAddressTypes) {
 			this.resolvedAddressTypes = Objects.requireNonNull(resolvedAddressTypes);
+			return this;
+		}
+
+		@Override
+		public NameResolverSpec retryTcpOnTimeout(boolean enable) {
+			this.retryTcpOnTimeout = enable;
 			return this;
 		}
 

--- a/reactor-netty-core/src/test/java/reactor/netty/transport/NameResolverProviderTest.java
+++ b/reactor-netty-core/src/test/java/reactor/netty/transport/NameResolverProviderTest.java
@@ -274,6 +274,14 @@ class NameResolverProviderTest {
 	}
 
 	@Test
+	void retryTcpOnTimeout() {
+		assertThat(builder.build().isRetryTcpOnTimeout()).isFalse();
+
+		builder.retryTcpOnTimeout(true);
+		assertThat(builder.build().isRetryTcpOnTimeout()).isTrue();
+	}
+
+	@Test
 	void roundRobinSelection() {
 		assertThat(builder.build().isRoundRobinSelection()).isFalse();
 

